File: pages/refactorings/extract-method/02-running-balance/After.scala
// Closing balance after a run of transactions; `step` is extracted from
// the fold lambda and stays nested, capturing `limit` and `fee`.
object After:
  def settle(opening: Int, limit: Int, fee: Int, txs: List[Int]): Int =
    def step(balance: Int, tx: Int): Int =
      val next = balance + tx
      if tx < 0 && next < -limit then next - fee else next
    txs.foldLeft(opening)(step)

File: pages/refactorings/extract-method/02-running-balance/After.hs
-- Closing balance after a run of transactions; each debit that leaves the
-- account below its overdraft limit is charged a fee.
module After where

import Data.List (foldl')

settle :: Int -> Int -> Int -> [Int] -> Int
settle opening limit fee = foldl' (step limit fee) opening

step :: Int -> Int -> Int -> Int -> Int
step limit fee balance tx =
  let next = balance + tx
  in if tx < 0 && next < negate limit then next - fee else next
